### PR TITLE
Add custom builders support.

### PR DESCRIPTION
**Pull request by wRAR**

### CODE REVIEW

**Comment by wRAR** [comment]
Not tested with scrapy-poet yet.

File: README.rst [hunk 1]
```diff
@@ -416,6 +416,47 @@ example:
...
+
+The build code also needs to know how to build ``Wheels`` instances:
+
+.. code-block:: python
+
+    def build(plan):
+        instances = {}
+        for fn_or_cls, kwargs_spec in plan:
+            if fn_or_cls in custom_builders:
+                instances[fn_or_cls] = custom_builders[cls](**kwargs_spec.kwargs(instances))
```
Comment (Gallaecio) [inline README.rst:455]:
```suggestion
                instances[fn_or_cls] = custom_builders[fn_or_cls](**kwargs_spec.kwargs(instances))
```

**Review by Gallaecio — COMMENTED** [review]
Looks great.

But I wonder how we will handle the case where [a wheel builder requires wheels (built differently) as input](https://zyte-spider-templates.readthedocs.io/en/latest/customization/page-objects.html#overriding-parsing).

**Comment by Gallaecio** [comment]
For clarification, this would be a test for what I mean:

```python
def test_plan_custom_builder_previous_dependency_change():
    class Item:
        def __init__(self, field: int):
            self.field = field

    class InitialPage:
        def __init__(self, b: B):
            pass

        def to_item(self):
            return Item(21)

    class FinalPage:
        def __init__(self, b: B, item: Item):
            self.item = item

        def to_item(self):
            item.field *= 2
            return item

    def fn(item: Item) -> int:
        return item.field + 1

    def initial_builder(page: InitialPage) -> Item:
        return page.to_item()

    def final_builder(page: FinalPage) -> Item:
        return page.to_item()

    # TODO: Maybe we need to make the value a list, or reverse keys and values?
    custom_builders = {
        Item: item_factory,
    }

    plan = andi.plan(fn, is_injectable={B, InitialPage, FinalPage}, custom_builders=custom_builders)
    assert plan == [
        (B, {}),
        (InitialPage, {"b": B}),
        (Item, {"page": InitialPage}),
        (Page, {"b": B, "item": Item}),
        (Item, {"FinalPage": FinalPage}),
        (fn, {"item": Item})
    ]
    instances = build(plan, custom_builders=custom_builders)
    assert set(instances.keys()) == {B, Page, Item, fn}
    assert instances[fn] == 43
```

**Comment by wRAR** [comment]
Looks like passing a static dict of builders to andi is not helpful for scrapy-poet, because it doesn't have a static mapping of  items to pages but can only ask `web_poet.RulesRegistry` for a page for a given item. So we need to either extend web-poet to export such a mapping for a given URL from the registry, or change the implementation of this PR to take a callback instead of (or in addition to?) a mapping. I'm also not sure yet how would that work for the item overrides use case.

cc @kmike

File: README.rst [hunk 2]
```diff
@@ -416,6 +416,53 @@ example:
...
+for an object built with a custom builder uses an instance of the ``andi.CustomBuilder``
+wrapper that contains the type to be built in the ``result_class_or_fn`` attribute and
+the callable for building it in the ``factory`` attribute:
+
+.. code-block:: python
+
+    def build(plan):
+        instances = {}
+        for fn_or_cls, kwargs_spec in plan:
+            if isinstance(fn_or_cls, CustomBuilder):
```
Comment (kmike) [inline README.rst:460]: An import would be helpful in this example

File: tests/test_plan.py [hunk 3]
```diff
@@ -498,3 +498,112 @@ def __call__(self,
...
+    custom_builders = get_custom_builders(Page)
+    plan = andi.plan(
+        fn,
+        is_injectable={B, Page},
+        externally_provided={Item},
+        custom_builder_fn=custom_builders.get
+    )
+    assert plan == [
+        (B, {}),
+        (Item, {}),
```
Comment (wRAR) [inline tests/test_plan.py:576]: This randomly fails on 3.5 because these two entries are sometimes swapped.

Comment (wRAR) [reply tests/test_plan.py:576]: Looks like the underlying issue is the iteration order of various dicts representing args, which on 3.5 is random between runs.

File: andi/andi.py [hunk 4]
```diff
@@ -307,51 +313,87 @@ class we want to override. If ``recursive_overrides`` is True, then
...
                     dependency_stack=None,
                     overrides=overrides,
-                    recursive_overrides=recursive_overrides)
+                    recursive_overrides=recursive_overrides,
+                    custom_builder_fn=custom_builder_fn,
+                    )
     return plan
 
 
+CustomBuilder = namedtuple("CustomBuilder", ["result_class_or_fn", "factory"])
```
Comment (kmike) [inline andi/andi.py:322]: What do you think about using a dataclass instead?

Comment (wRAR) [reply andi/andi.py:322]: No 3.5 support

File: andi/andi.py [hunk 5]
```diff
@@ -307,51 +313,87 @@ class we want to override. If ``recursive_overrides`` is True, then
...
 
-    # At this point the class/function must be injectable for non root cases
-    assert is_root_call or is_injectable(strip_annotated(class_or_func))
+    if not custom_builder_result:
+        plan_key = class_or_func  # type: Union[Callable, CustomBuilder]
+    else:
+        plan_key = CustomBuilder(custom_builder_result, class_or_func)
+
+    # At this point the class/function must be injectable or built by a custom builder for non root cases
+    assert is_root_call or is_injectable(strip_annotated(class_or_func)) or custom_builder_result
```
Comment (kmike) [inline andi/andi.py:349]: very minor: would it make sense to check custom_builder_result first, to save a couple of function calls if it's true?

File: andi/andi.py (hunk 4 above)
Comment (kmike) [reply andi/andi.py:322]: I think we can drop 3.5

Comment (kmike) [reply andi/andi.py:322]: And probably 3.6 and 3.7 as well, as we did in other packages

**Review by kmike — APPROVED** [review]

File: andi/andi.py (hunk 4 above)
Comment (wRAR) [reply andi/andi.py]: Done

**Pull request merged**